src/relax/transform/lift_transform_params.cc

```diff
@@ -34,136 +34,211 @@
-    return {f_transform_params, bindings_, output_to_index_};
+  Expr VisitExpr_(const VarNode* var) final {
+    if (auto it = input_remap_.find(var); it != input_remap_.end()) {
+      return builder_->Emit((*it).second);
+    } else {
+      return ExprMutator::VisitExpr_(var);
+    }
   }
 
-  void VisitBinding_(const VarBindingNode* binding, const CallNode* val) final {
-    static const Op& call_tir_op_ = Op::Get("relax.call_tir");
-    if (val->op.same_as(call_tir_op_)) {
-      if (std::all_of(val->args.begin(), val->args.end(), [&](const Expr& arg) {
-            const auto* var = arg.as<VarNode>();
-            return !var || bindings_.count(GetRef<Var>(var));
-          })) {
-        bindings_.insert(binding->var);
-        output_to_index_[binding->var] = outputs.size();
-
-        auto new_var = builder_->Emit(Substitute(binding->value, var_remap_));
-        // TODO(wuwei): we can prune the outputs and only keep the final result of transformations
-        // when there are a sequence of transformations applied to the parameters.
-        outputs.push_back(new_var);
-      }
+  // The input parameters of the function.
+  Array<Var> inputs_;
+  // Remap from the original input variable to TupleGetItem from the packed parameter tuple, which
+  // is the input of the lifted function.
+  std::unordered_map<const VarNode*, Expr> input_remap_;
+  // The bindings that are lifted.
+  Array<VarBinding> bindings_;
+  // The variables that are marked as the output of the function.
+  std::unordered_set<Var, ObjectPtrHash, ObjectPtrEqual> outputs_;
+};
+
+/*!
+ * \brief Visitor that creates the plan of lifting transform params.
+ *
+ * Starting from the parameters of the function (they are the initial set of lifted bindings), we
+ * will visit the body of the function to find the bindings that can be lifted. A binding can be
+ * lifted if all the variables that it depends on are also lifted.
+ *
+ * When a binding cannot be lifted, all the variables that 1) it depends on, and 2) have been
+ * lifted, will be marked as the boundary variable and will be in the output of the lifted function.
+ */
+class LiftTransformParamsPlanner : public ExprVisitor {
+ public:
+  LiftTransformParamsInfoPlan Plan(const Function& function, int num_inputs) {
+    for (int i = num_inputs; i < static_cast<int>(function->params.size()); ++i) {
+      builder_.AddInput(function->params[i]);
+      lifted_bindings_.emplace(function->params[i]);
     }
+    VisitExpr(function->body);
+
+    const auto& [f_transform_params, output_to_index] = builder_.Build();
+    return {f_transform_params, output_to_index, std::move(lifted_bindings_)};
   }
 
  private:
-  std::unordered_set<Var, ObjectHash, ObjectEqual> bindings_;
-  BlockBuilder builder_;
+  void VisitBinding_(const VarBindingNode* binding) final {
+    std::vector<const VarNode*> producers;
+    bool can_lift = true;
+
+    PostOrderVisit(binding->value, [&](const ObjectRef& obj) {
```

Comment:
binding->value may contain Tuple.
Also, we may want to make sure that we are working in the context of a DataflowBlock.
Once some param to lift is used outside of df block, we might want to reject such programs.

Reply:
Also, do we consider control flow in the relax program?
For example
```py
if pred:
   x0 = transpose([1, 2], param)
else:
   x0 = transpose([0, 1], param)
```

Reply:
control flow is not supported, I added a test case

Reply:
In this case I think we need to be cautious about IfNode and not recurse into it, since it is possible for a IfNode to be
```python
if cond:
  with R.dataflow():
    ...
    R.output(..)
  ...
else:
  ...
```

Reply:
@MasterJH5574  I think we care recurse into this case.
```
if cond:
  with R.dataflow():
    w1 = convert_layout(w)
    ...
    R.output(..)
  ...
else:
  ...
```
it will lift the transform out of if condition, it becomes
```
def transform_params(w):
  w1 = convert_layout(w)
  return (w1,)

def main(x, params):
  if cond:
    w1 = params[0]
    ...
  else:
    ...
```

Reply:
Yeah sorry I didn’t wrote out the example completely. I meant the cases where both “then” and “else” contain dataflow blocks (by wrapping Bohan’s example with dataflow blocks).
```python
if cond:
  with R.dataflow():
    ...
    R.output(..)
  ...
else:
  with R.dataflow():
    ...
    R.output(...)
  ...
```

Reply:
In this case, transformation in both branches will be lifted out. In `transform_params` we will compute the transformation in both branches (it's can be redundant since we only need one of them during inference, but having both of them pre-computed, we will save computation during runtime).

Reply:
I see, as long as we can pre-compute then it is fine. Thanks for the elaboration!